### PR TITLE
Code review for Crociani and De Zen

### PR DESCRIPTION
**Pull request by tonin-alessandra**
Hi guys, this is my code to review.
It is about Kademlia's dictionary and, in particular, you have to review these parts:
- interface PingListener
- class SMSNetworkListener
- class SMSNetworkManager: methods join, onJoinAgreedReply, onJoinProposal, ping, 
   onPingRequest, onPingEchoReply, refreshBucket, republishKeys
- class SMSDistributedNetworkDictionary:  method addUser

I added a comment "CODE REVIEW FOR CROCIANI AND DE ZEN" in each class, but anyway, if you have any problem or question please ask me.
Thank you, and good review!

### CODE REVIEW

File: networklibrary/src/main/java/com/eis/communication/network/PingListener.java [hunk 1]
```diff
@@ -1 +1,16 @@
...
+
+/**
+ * A listener waiting for a {@code PING} request
+ *
+ * @author Alessandra Tonin
+ *
+ * CODE REVIEW FOR CROCIANI AND DE ZEN
+ */
+public interface PingListener {
+    void onPingReply(SMSPeer peer);
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/communication/network/PingListener.java]: Specs could be added to methods in order to better explain what the interface does.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java [hunk 2]
```diff
@@ -34,15 +40,53 @@ public SMSDistributedNetworkDictionary(SMSKADPeer mySelf) {
...
+     */
+    @Override
+    public void addUser(final SMSKADPeer newUser) {
+
+        //bucketIndex is the closer bucket of mySelf containing newUser
+        int bucketIndex = getBucketContaining(newUser.getNetworkAddress());
+        //If it's actually the current user we don't add himself
+        if (bucketIndex == -1) return;
 
+        if (buckets[bucketIndex] == null)
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: I like this use of lazy initialization 👍

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java [hunk 3]
```diff
@@ -34,15 +40,53 @@ public SMSDistributedNetworkDictionary(SMSKADPeer mySelf) {
...
+            buckets[bucketIndex] = new ArrayList<>();
+
+        final ArrayList<SMSKADPeer> usersInBucket = buckets[bucketIndex];
+
+        if (usersInBucket.size() == KADEMLIA_K) {
+            PingListener listener = new PingListener() {
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: If you declare this listener as a private class you can reuse it if you ever need to, I guess you only need it here right now, but it could be different in the future, also it helps focusing the attention on the core of the method instead of the ping. Good job implementing it inside the Dictionary instead of having the Dictionary itself a listener, this way you are preventing unwanted pings 👍

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java [hunk 4]
```diff
@@ -34,15 +40,53 @@ public SMSDistributedNetworkDictionary(SMSKADPeer mySelf) {
...
+                @Override
+                public void onPingTimedOut(SMSPeer peer) {
+                    usersInBucket.remove(peer); //peer does not respond, so we remove it
+                    usersInBucket.add(newUser); //instead we add the new user at the tail
+                }
+            };
+            SMSKADPeer head = usersInBucket.get(0);
+            SMSNetworkManager.getInstance().ping(head, listener);
+        } else {
+            usersInBucket.remove(newUser); //does nothing if not present
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: This is unrelated to your code, just a suggestion, if you switch over to Sets to represent buckets (or if you extend it in a Bucket class) you can avoid trying to remove nodes before adding them.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkListener.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    /**
+     * Checks if the received message is a command for the kad dictionary, checks which command
+     * has been received and calls processReply or processRequest.
+     *
+     * @param message the received message
+     */
+    @Override
+    public void onMessageReceived(SMSMessage message) {
+        String[] splitMessageContent = message.getData().split(SPLIT_CHAR);
+        String messagePrefix = splitMessageContent[0];
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkListener.java]: If some day you were to move around the order of content in the message, assigning the index values to a constant can prove useful.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkListener.java [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        int parts = splitMessageContent.length;
+        //Check if it's a reply
+        for (SMSNetworkManager.ReplyType replyCommand : SMSNetworkManager.ReplyType.values()) {
+            if (replyCommand.toString().equals(messagePrefix)) {
+                processReply(replyCommand, message.getPeer(), parts == 1 ? "" : splitMessageContent[1]);
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkListener.java]: Same as above for the index, you can also assign the empty string to a field.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java [hunk 7]
```diff
@@ -33,11 +31,54 @@
  * @author Alessandra Tonin
  * @author Alberto Ursino
  * @author Luca Crema
+ *
+ * CODE REVIEW FOR CROCIANI AND DE ZEN
  */
 @SuppressWarnings("WeakerAccess")
 public class SMSNetworkManager implements NetworkManager<SMSKADPeer, SerializableObject, SerializableObject, KADInvitation> {
+    //Requests and replies: for further details, check out SMSNetworkListener
+    enum RequestType {
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: Another suggestion unrelated to your code. In my opinion you guys could add a short code as a parameter for the enum, you would save some bytes compared to writing the enum as a String. It's fairly simple, a good example for that is here:
https://www.baeldung.com/java-enum-values#adding-constructor

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java [hunk 8]
```diff
@@ -97,10 +141,64 @@ public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
         });
     }
 
+    /**
+     * Join the network
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: Could be a bit more verbose than just "Join the network".

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java [hunk 9]
```diff
@@ -97,10 +141,64 @@ public void onSMSSent(SMSMessage message, SMSMessage.SentState sentState) {
...
+     *
+     * @param invitation The invitation message
+     * @author Alessandra Tonin, Marco Mariotto
+     */
+    @Override
+    public void join(KADInvitation invitation) {
+        if (invitation.getGuest() != mySelf)
+            throw new IllegalArgumentException("The invitation is not valid: it is intended for another user");
+        findClosestNodes(mySelf.networkAddress, new FindNodeListener<SMSKADPeer>() {
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: Same as the listener in the Dictionary

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java [hunk 10]
```diff
@@ -377,10 +474,55 @@ synchronized protected void onValueFoundReply(String replyContent) {
...
+     */
+    synchronized public void ping(SMSPeer peer, PingListener listener) {
+        SMSCommandMapper.sendRequest(RequestType.PING, peer);
+        listenerHandler.registerPingListener(peer, listener);
+    }
 
+    /**
+     * Method called when a {@link RequestType#PING} request has been received. Sends a {@link ReplyType#PING_ECHO) command back.
+     *
+     * @param peer who requested a ping
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: It would be better (albeit it might sound redundant) to write:
`@param peer The Peer who requested a ping`
Since when Javadocs are created the name of the parameter isn't necessarily part of the commented sentence.
For example here:
https://docs.oracle.com/javase/8/docs/api/java/util/ArrayList.html#add-int-E-
They write "element - element to be inserted" instead of "element - to be inserted".

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java [hunk 11]
```diff
@@ -397,5 +539,24 @@ private void updateLastLookup(KADAddress address) {
...
 
-    
+    /**
+     * Republishes all keys of the local dictionary. Called by the RepublishService only every {@link #KADEMLIA_REPUBLISH_PERIOD_MILLIS} milliseconds.
+     *
+     * @author Alessandra Tonin, Marco Mariotto
+     */
+    synchronized public void republishKeys() {
+        for (final KADAddress resourceKey : dict.getKeys()) {
+            FindNodeListener listener = new FindNodeListener() {
```
Comment (RicDeZen) [inline networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: Same as the other listeners

**Comment by RicDeZen** [comment]
- Code is well formatted.
- Names have a clear meaning.
- Naming conventions are kept consistent.
- No TODOs or @fixme.
- There is an instance of Lazy Initialization which, albeit simple, I liked quite a bit. Listener design pattern has been used as local class a couple of times, for better reusability and readability it could be moved to a private class instead.
- Except where marked (local listeners) the code is very well readable.
- There are some instances of something that may or may not be considered hardwiring, regarding array indexes for arrays with a fixed structure, and some empty strings.
- Some specs could be written better, I commented on the specific points where I think this is the case.
I added two suggestions regarding the architecture of the code which are not strictly related to your code, but I found no author tag.

Overall the code is fairly good. Nothing in particular I would change, I just commented on some small fixes that could be applied. As I said some of the specs and comments are a bit lacking, for this you don't need to look far, most specs are already very good, I'd say the ones in SMSNetworkListener were the best.

**Comment by riccardo-crociani** [comment]
**PingListener**
You should add the specs to the methods onPingReply and onPingTimedOut

**onJoinAgreedReply**:

- line 180: that is an interesting idea

**SMSNetworkManager**:

- line 145: "Join the network"

- line 170: "Method called when... "

I think you should be a little more accurate and explain what the methods do

**General comment**

- The code is quite well formatted

- There are no TODOs/ hardwiring

- Well organized

Overall the code seems to be quite good and to work properly. Maybe some specs could have been written with more accuracy but in general, the code is easy to read. Good job!

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java (hunk 2 above)
Comment (tonin-alessandra) [reply networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: Thank you.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java (hunk 4 above)
Comment (tonin-alessandra) [reply networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: Thank you for the suggestion, I'll talk about this with my group.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java (hunk 10 above)
Comment (tonin-alessandra) [reply networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: You're right, thanks.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java (hunk 3 above)
Comment (tonin-alessandra) [reply networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSDistributedNetworkDictionary.java]: Thank you for this suggestion.
As you said, I only need this listener here, so it should not be a problem to keep like this for the moment, but you're also right saying in the future I might need to reuse it: I'll think about it and maybe I'll change.

File: networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java (hunk 7 above)
Comment (tonin-alessandra) [reply networklibrary/src/main/java/com/eis/networklibrary/kademlia/SMSNetworkManager.java]: We already thought about this, but we had no time to change everything. We'll try to fix it, thanks.

**Comment by tonin-alessandra** [comment]
Thanks to both of you for the reviews, your suggestions made my code better!
